js/main.js

```diff
@@ -0,0 +1,263 @@
+    If it isn't, then proceed
+  Set up a request for the 0th item in objectIDs
+  Here's what should happen when it's done loading
+    Get the response and store it in metArtObj
+    If it has an image URL
+      (In the future, save the info about the painting)
+      We're done, return the URL and exit
+    If it doesn't have a URL
+      Set the index from the 0th item to the 1st item
+      Call this function again with the new index
+  Now send it and let's see what happens
+*/
+function getArtwork(isStart) {
+  // This function gets artwork and puts it into the artObjCache if it has an accessible URL
+  if (artObjCache.length >= cacheItemsNum) {
+    // Don't keep sending requests if the cache is full, prevent button spamming to abuse API
+    return;
+  }
+
+  var randDept = Math.floor(Math.random() * metDepts.length);
+  // Department id's are not continuous, some are missing (from 1 - 21, 2 and 20 are missing). Access by index
+  var searchRequest = metSearch(metDepts[randDept].departmentId);
+  var searchResultsIdx = 0; // this is declared outside of the search results, making it suitable for iterating through search results
+
+  function handleSearchResponse() {
+    metSearchResults = searchRequest.response;
+    metSearchResults.objectIDs = shuffleArray(metSearchResults.objectIDs);
+    if (metSearchResults.objectIDs === null) {
+      // if results were bad, exit for now
+      return;
+    } else if (searchResultsIdx >= metSearchResults.objectIDs.length) {
+      // if we exhausted the list of id's, pull again
+      getArtwork();
+    }
+
+    var currentObjId = metSearchResults.objectIDs[searchResultsIdx];
+    // if this artwork been shown already, skip it
+    while (data.shownObjectIds.includes(currentObjId)) {
+      searchResultsIdx++;
+      currentObjId = metSearchResults.objectIDs[searchResultsIdx];
+    }
+
+    // Now that we have a new, never before seen currentObjId, we can attempt to acquire it.
+    var acquireRequest = metAcquireArt(currentObjId);
+    // acquireRequest.onload = function () {
+    acquireRequest.addEventListener('load', function handleAcquireReponse() {
+      metArtObj = acquireRequest.response;
+      // Store the acquired object ID so we know we've seen it already
+      data.shownObjectIds.push(metArtObj.objectID);
+      if (metArtObj.primaryImage !== '') {
+        if (isStart) {
+          // If this is the first time running, go straight to showing it rather than caching it
+          setImage(metArtObj);
+        } else {
+          artObjCache.push(metArtObj);
+        }
+      } else {
+        searchResultsIdx++;
+        handleSearchResponse(); // This will use the new value of searchResultsIdx since handleSearchResponse calls searchResultsIdx from a higher scope than itself
+      }
+      acquireRequest.removeEventListener('load', handleAcquireReponse);
+    });
+    acquireRequest.send();
+    searchRequest.removeEventListener('load', handleSearchResponse);
+  }
```

Comment:
Is there any particular reason this needs to be inside of another function definition?

Reply:
Per our Slack discussion on variables (and their attached event listeners) being garbage collected, inner functions have been declared outside.